hwcomposer/hwcomposer_backend_v11.cpp

```diff
@@ -86,97 +94,412 @@ static void hwc11_callback_invalidate(const struct hwc_procs *)
+        HWC_SCREEN_FRAMEBUFFER_TARGET_LAYER = 1,
 
-    int retireFenceFd = -1;
+        // Number of layers we need to allocate space for
+        HWC_SCREEN_REQUIRED_LAYERS = 2,
+    };
+
+    HwComposerScreen_v11(hwc_composer_device_1_t *hwc_device, int id)
+        : hwc_device(hwc_device)
+        , id(id)
+        , hwc_list(nullptr)
+        , m_screen_width(0)
+        , m_screen_height(0)
+    {
+        size_t needed_size = sizeof(hwc_display_contents_1_t) +
+            HWC_SCREEN_REQUIRED_LAYERS * sizeof(hwc_layer_1_t);
+
+        hwc_list = (hwc_display_contents_1_t *) calloc(1, needed_size);
+
+        // Need to set this here, and not every time in relayout
+        hwc_list->numHwLayers = 2;
+        hwc_list->retireFenceFd = -1;
+#ifdef HWC_DEVICE_API_VERSION_1_3
+        hwc_list->outbuf = 0;
+        hwc_list->outbufAcquireFenceFd = -1;
+#endif
+    }
+
+    void update_size() {
+        get_screen_size(hwc_device, id, &m_screen_width, &m_screen_height);
+    }
+
+    bool relayout(int width, int height)
+    {
+        // Source rectangle of the desktop
+        const hwc_rect_t source_rect = {
+            0, 0, width, height
+        };
+
+        // Don't call getDisplayAttributes too often here since it can cause big slowdowns.
+        if (m_screen_width <= 0 || m_screen_height <= 0) {
+            update_size();
+        }
+
+        // Destination rectangle on the actual screen
+        const hwc_rect_t dest_rect = {
+            0, 0, m_screen_width, m_screen_height
+        };
+
+        hwc_layer_1_t *layer = NULL;
+
+        layer = getLayer(HWC_SCREEN_FRAMEBUFFER_LAYER);
+        resetLayer(layer);
+
+        layer->compositionType = HWC_FRAMEBUFFER;
+        layer->hints = 0;
+        layer->flags = 0;
+        layer->handle = 0;
+        layer->transform = 0;
+        layer->blending = HWC_BLENDING_NONE;
+    #ifdef HWC_DEVICE_API_VERSION_1_3
+        layer->sourceCropf.top = 0.0f;
+        layer->sourceCropf.left = 0.0f;
+        layer->sourceCropf.bottom = (float) height;
+        layer->sourceCropf.right = (float) width;
+    #else
+        layer->sourceCrop = source_rect;
+    #endif
+        layer->displayFrame = dest_rect;
+        layer->visibleRegionScreen.numRects = 1;
+        layer->visibleRegionScreen.rects = &layer->displayFrame;
+        layer->acquireFenceFd = -1;
+        layer->releaseFenceFd = -1;
+    #if (ANDROID_VERSION_MAJOR >= 4) && (ANDROID_VERSION_MINOR >= 3) || (ANDROID_VERSION_MAJOR >= 5)
+        // We've observed that qualcomm chipsets enters into compositionType == 6
+        // (HWC_BLIT), an undocumented composition type which gives us rendering
+        // glitches and warnings in logcat. By setting the planarAlpha to non-
+        // opaque, we attempt to force the HWC into using HWC_FRAMEBUFFER for this
+        // layer so the HWC_FRAMEBUFFER_TARGET layer actually gets used.
+        bool tryToForceGLES = !qgetenv("QPA_HWC_FORCE_GLES").isEmpty();
```

Comment:
What do you think about static bool instead calling qgetenv on every frame.
```suggestion
        static bool tryToForceGLES = qEnvironmentVariableIsSet("QPA_HWC_FORCE_GLES");
```